engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java

```diff
@@ -0,0 +1,769 @@
+            // Check if templates path exists
+            try {
+                Set<String> hypervisors = hypervisorsInUse.stream().map(Enum::name).
+                        map(name -> name.toLowerCase(Locale.ROOT)).map(SystemVmTemplateRegistration::getHypervisorName).collect(Collectors.toSet());
+                List<String> templates = new ArrayList<>();
+                for (Hypervisor.HypervisorType hypervisorType : hypervisorsInUse) {
+                    templates.add(fileNames.get(hypervisorType));
+                }
+
+                boolean templatesFound = true;
+                for (String hypervisor : hypervisors) {
+                    String matchedTemplate = templates.stream().filter(x -> x.contains(hypervisor)).findAny().orElse(null);
+                    if (matchedTemplate == null) {
+                        templatesFound = false;
+                        break;
+                    }
+                    MessageDigest mdigest = MessageDigest.getInstance(hashAlgorithm);
+                    File tempFile = new File(templatesPath + matchedTemplate);
+                    String templateChecksum = calculateChecksum(mdigest, tempFile);
+                    if (!templateChecksum.equals(newTemplateChecksum.get(getHypervisorType(hypervisor)))) {
+                        LOGGER.error(String.format("Checksum mismatch: %s != %s ", templateChecksum, newTemplateChecksum.get(getHypervisorType(hypervisor))));
+                        templatesFound = false;
+                        break;
+                    }
+                }
+
+                if (!templatesFound) {
+                    LOGGER.info("SystemVm template not found. Cannot upgrade system Vms");
+                    throw new CloudRuntimeException("SystemVm template not found. Cannot upgrade system Vms");
+                }
+
+                // Perform Registration if templates not already registered
+                List<Long> zoneIds = getEligibleZoneIds(conn);
+                for (Long zoneId : zoneIds) {
+                    Pair<String, Long> storeUrlAndId = getNfsStoreInZone(conn, zoneId);
+                    List<String> hypervisorList = fetchAllHypervisors(conn, zoneId);
+                    for (String hypervisor : hypervisorList) {
+                        Hypervisor.HypervisorType name = Hypervisor.HypervisorType.getType(hypervisor);
+                        String templateName = NewTemplateNameList.get(name);
+                        Pair<Hypervisor.HypervisorType, String> hypervisorAndTemplateName = new Pair<Hypervisor.HypervisorType, String>(name, templateName);
+                        long templateId = isTemplateAlreadyRegistered(conn, hypervisorAndTemplateName);
+                        if (templateId != -1) {
+                            continue;
+                        }
+                        registerTemplate(conn, hypervisorAndTemplateName, storeUrlAndId);
+                    }
+                    unmountStore();
+                }
```

Comment:
There are several process here, we could extract some to minor methods. Also, we could add unit tests to these methods.